### ISSUE TITLE
sbt/codegen: multiple input dirs? (`sourceDirectory` vs `sourceDirectories`)

### ISSUE THREAD

**Issue by kubukoz** [disneystreaming/smithy4s#464]
An sbt project doesn't necessarily have exactly one source directory. For example, in an sbt-typelevel project with `CrossType.Full`, it makes a huge difference:

<img width="1375" alt="image" src="https://user-images.githubusercontent.com/894884/190934611-2793a7a3-1ac0-4cfe-9c37-c85cced77173.png">

Apparently, in this setup, if I have both JS and JVM enabled, I need to copy-paste (or symlink, at least) the Smithy files so that they are seen in each variant's main `sourceDirectory`.

A workaround is setting `smithy4sInputDir` explicitly, but I feel like I'm missing something - or that it would be a good idea to support multiple `sourceDirectories` entries in the plugin itself.

**Comment by daddykotex** [comment]
I have also been _blocked_ by this. I can't remember exactly what was my use case. I think I had extracted to specs from a jar and I was required to copy paste them in.

**Comment by Baccata** [comment]
I'm happy to review a PR that would add this, in a way that would deprecate the current setting in favour of a new one

**Comment by lewisjkl** [comment]
Was just talking with @daddykotex and I think there are two different issues here, the main one being that @kubukoz was needing to duplicate smithy specs across folders in a project that used sbt-crossproject. Imo the solution there should be to update the configuration in the project to look at a single source directory e.g. 

```scala
(Compile / smithy4sInputDir) := (CrossType.Pure
      .sharedResourcesDir(baseDirectory.value, "main")
      .get / "smithy")
```

That way you can have a single location for smithy sources and it will be used across your different cross variants.

As far as using multiple "sourceDirectories," I don't see how that solves this particular issue. To me, that just enables you to duplicate the specs to multiple locations and have them found when using sbt-crossproject which is doing the opposite of what we want. There are other (hypothetical) uses cases I can think of where multiple source directories would be useful, but that seems to be tangential to this issue. However, I might be missing something so let me know!

**Comment by daddykotex** [comment]
So we spend a bit of time discussing this and I want to be clear about the issue I encountered because I think we did not have the same issue. Basically, when I said: 
> I have also been blocked by this. I can't remember exactly what was my use case. I think I had extracted to specs from a jar and I was required to copy paste them in.

I meant: I have a project, with one smithy file in it, in the default directory and it works fine. Now I'd like to try something and replace my dependency on aws-traits (for example) and instead use the content of directory **X** (on top of the default one). And in directory **X**, I have extracted the smithy files for the aws-traits jar and I'm just amending the definitions for whatever reason.

I think this is really different from your use case and my use case (supporting different sourceDirectory) does not warrant a new feature or a change to the current implementation.

**Comment by kubukoz** [comment]
> As far as using multiple "sourceDirectories," I don't see how that solves this particular issue. To me, that just enables you to duplicate the specs to multiple locations and have them found when using sbt-crossproject which is doing the opposite of what we want. 

It solves this issue while also supporting platform/version-specific Smithy files. Each project in the crossproject setup gets a bunch of source directories by default, including the shared one but also its own ones, so you could put the shared specs into the shared dir and some more specific ones into the... not shared dir.

**Comment by lewisjkl** [comment]
> It solves this issue while also supporting platform/version-specific Smithy files. Each project in the crossproject setup gets a bunch of source directories by default, including the shared one but also its own ones, so you could put the shared specs into the shared dir and some more specific ones into the... not shared dir.

I think I am starting to see what you are saying, but it isn't quite clicking for me. So we would configure multiple source directories instead of `smithy4sInputDir` it would be `smithy4sInputDirs` and have multiple? Or if we automatically added multiple source directories, how would we go about that?

**Comment by kubukoz** [comment]
>  So we would configure multiple source directories instead of smithy4sInputDir it would be smithy4sInputDirs and have multiple?

yeah

> Or if we automatically added multiple source directories, how would we go about that?

We _might_ be able to just replace `sourceDirectory` in the default:

https://github.com/disneystreaming/smithy4s/blob/3f965eff77eceb7add9525434759fedc2bae155f/modules/codegen-plugin/src/smithy4s/codegen/Smithy4sCodegenPlugin.scala#L101

with something like `config / smithy4sInputDirs := (config / sourceDirectories).value.map(_ / "smithy")`.

Maybe I misunderstood the question though.

**Comment by lewisjkl** [comment]
Ah I think I see where my misunderstanding was. For some reason I was thinking that using sourceDirectories worked a bit differently than that. I will give it a go and see what I can come up with.

**Comment by kubukoz** [comment]
hmm, I think we want to use `unmanagedSourceDirectories` instead: `sourceDirectories` includes the managed ones.

**Comment by daddykotex** [comment]
Ok, you build upon `unmanagedSourceDirectories`, so if a plugin like cross-project already sets this to a bunch a variant, then you can use it to do the same for smithy

I see

**Comment by kubukoz** [comment]
closed in #587
